processor/transformprocessor/internal/metrics/func_drop_bucket.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlmetric"
+)
+
+type DropBucketArguments struct {
+	Pattern string
+}
+
+func newDropBucketFactory() ottl.Factory[ottlmetric.TransformContext] {
+	return ottl.NewFactory(
+		"drop_bucket",
+		&DropBucketArguments{},
+		createDropBucketFunction,
+	)
+}
+
+func createDropBucketFunction(_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	args, ok := oArgs.(*DropBucketArguments)
+
+	if !ok {
+		return nil, errors.New("DropBucketFactory args must be of type *DropBucketArguments[K]")
+	}
+
+	return dropBucketFunc(args.Pattern), nil
+}
+
+func dropBucketFunc(pattern string) ottl.ExprFunc[ottlmetric.TransformContext] {
+	return func(_ context.Context, tCtx ottlmetric.TransformContext) (any, error) {
+		compiledPattern, err := regexp.Compile(pattern)
+		if err != nil {
+			return nil, fmt.Errorf("the regex pattern supplied to drop_bucket is not a valid pattern: %w", err)
+		}
+
+		metric := tCtx.GetMetric()
+
+		switch metric.Type() {
+		case pmetric.MetricTypeHistogram:
+			dropBucketHistogram(&metric, compiledPattern)
+		case pmetric.MetricTypeGauge:
+			dropBucketGauge(&metric, compiledPattern)
+		case pmetric.MetricTypeSum:
+			dropBucketSum(&metric, compiledPattern)
+		}
+		return nil, nil
+	}
+}
+
+func dropBucketHistogram(metric *pmetric.Metric, compiledPattern *regexp.Regexp) {
+	dps := metric.Histogram().DataPoints()
+	for i := 0; i < dps.Len(); i++ {
+		hdp := dps.At(i)
+		bounds := hdp.ExplicitBounds()
+		counts := hdp.BucketCounts()
+
+		var newBounds []float64
+		var newCounts []uint64
+		var removedBoundIndices []int
+
+		for bIdx := 0; bIdx < bounds.Len(); bIdx++ {
+			if compiledPattern.MatchString(strconv.FormatFloat(bounds.At(bIdx), 'f', -1, 64)) {
+				removedBoundIndices = append(removedBoundIndices, bIdx)
+			} else {
+				newBounds = append(newBounds, bounds.At(bIdx))
+			}
+		}
+
+		if len(removedBoundIndices) == 0 {
+			continue
+		}
+
+		countsToRemove := make(map[int]bool)
+		for _, bIdx := range removedBoundIndices {
+			countsToRemove[bIdx+1] = true
+		}
+
+		for cIdx := 0; cIdx < counts.Len(); cIdx++ {
+			if !countsToRemove[cIdx] {
+				newCounts = append(newCounts, counts.At(cIdx))
+			}
+		}
+
+		newBoundsSlice := pcommon.NewFloat64Slice()
+		newBoundsSlice.FromRaw(newBounds)
+		newBoundsSlice.MoveTo(hdp.ExplicitBounds())
+
+		newCountsSlice := pcommon.NewUInt64Slice()
+		newCountsSlice.FromRaw(newCounts)
+		newCountsSlice.MoveTo(hdp.BucketCounts())
+	}
+}
+
+func dropBucketGauge(metric *pmetric.Metric, compiledPattern *regexp.Regexp) {
+	dps := metric.Gauge().DataPoints()
+	dps.RemoveIf(func(dp pmetric.NumberDataPoint) bool {
+		attrs := dp.Attributes()
+		if le, ok := attrs.Get("le"); ok && le.Type() == pcommon.ValueTypeStr {
+			bound, err := strconv.ParseFloat(le.Str(), 64)
+			if err == nil && compiledPattern.MatchString(strconv.FormatFloat(bound, 'f', -1, 64)) {
+				return true
+			}
```

Comment:
why not directly using the regexp on the string value? Please comment why that is not the case.

Reply:
You are totally right.